pkg/operator/client/client.go

```diff
@@ -247,88 +247,148 @@ func (c *Client) deployManifests(deployArgs operatortypes.DeployAppArgs) (*deplo
+	if deployArgs.PreviousKotsKinds != nil && deployArgs.PreviousKotsKinds.V1Beta1HelmCharts != nil {
+		for _, kotsChart := range deployArgs.PreviousKotsKinds.V1Beta1HelmCharts.Items {
+			kc := kotsChart
+			prevKotsV1Beta1Charts = append(prevKotsV1Beta1Charts, &kc)
 		}
-		defer os.RemoveAll(tmpDir)
+	}
 
-		err = ioutil.WriteFile(path.Join(tmpDir, "archive.tar.gz"), deployArgs.PreviousCharts, 0644)
-		if err != nil {
-			return nil, errors.Wrap(err, "failed to write previous archive")
+	curV1Beta1KotsCharts := []kotsutil.HelmChartInterface{}
+	if deployArgs.KotsKinds != nil && deployArgs.KotsKinds.V1Beta1HelmCharts != nil {
+		for _, kotsChart := range deployArgs.KotsKinds.V1Beta1HelmCharts.Items {
+			kc := kotsChart
+			curV1Beta1KotsCharts = append(curV1Beta1KotsCharts, &kc)
 		}
+	}
 
-		prevHelmDir = path.Join(tmpDir, "prevhelm")
-		if err := os.MkdirAll(prevHelmDir, 0755); err != nil {
-			return nil, errors.Wrap(err, "failed to create dir to stage previous helm archive")
+	prevKotsV1Beta2Charts := []kotsutil.HelmChartInterface{}
+	if deployArgs.PreviousKotsKinds != nil && deployArgs.PreviousKotsKinds.V1Beta2HelmCharts != nil {
+		for _, kotsChart := range deployArgs.PreviousKotsKinds.V1Beta2HelmCharts.Items {
+			kc := kotsChart
+			prevKotsV1Beta2Charts = append(prevKotsV1Beta2Charts, &kc)
 		}
+	}
 
-		if err := tarGz.Unarchive(path.Join(tmpDir, "archive.tar.gz"), prevHelmDir); err != nil {
-			return nil, errors.Wrap(err, "falied to unarchive previous helm archive")
+	curV1Beta2KotsCharts := []kotsutil.HelmChartInterface{}
+	if deployArgs.KotsKinds != nil && deployArgs.KotsKinds.V1Beta2HelmCharts != nil {
+		for _, kotsChart := range deployArgs.KotsKinds.V1Beta2HelmCharts.Items {
+			kc := kotsChart
+			curV1Beta2KotsCharts = append(curV1Beta2KotsCharts, &kc)
 		}
 	}
 
-	var curHelmDir string
-	var installResult *commandResult
-	if len(deployArgs.Charts) > 0 {
-		tmpDir, err := ioutil.TempDir("", "helm")
-		if err != nil {
-			return nil, errors.Wrap(err, "failed to create temp dir to stage currently deployed archive")
+	removedCharts, err := getRemovedCharts(prevV1Beta1HelmDir, curV1Beta1HelmDir, prevKotsV1Beta1Charts, curV1Beta1KotsCharts, prevV1Beta2HelmDir, curV1Beta2HelmDir, prevKotsV1Beta2Charts, curV1Beta2KotsCharts)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to find removed charts")
+	}
+
+	// uninstall removed charts
+	if len(removedCharts) > 0 {
+		kotsCharts := append(prevKotsV1Beta1Charts, prevKotsV1Beta2Charts...)
+
+		v1Beta1ChartsDir := ""
+		if prevV1Beta1HelmDir != "" {
+			v1Beta1ChartsDir = filepath.Join(prevV1Beta1HelmDir, "charts")
+		}
+		v1Beta2ChartsDir := ""
+		if prevV1Beta2HelmDir != "" {
+			v1Beta2ChartsDir = filepath.Join(prevV1Beta2HelmDir, "helm")
 		}
-		defer os.RemoveAll(tmpDir)
 
-		err = ioutil.WriteFile(path.Join(tmpDir, "archive.tar.gz"), deployArgs.Charts, 0644)
+		err := c.uninstallWithHelm(v1Beta1ChartsDir, v1Beta2ChartsDir, kotsCharts)
```

Comment:
hmmm, shouldn't this be
```suggestion
		err := c.uninstallWithHelm(v1Beta1ChartsDir, v1Beta2ChartsDir, removedCharts)
```

i'm also looking at the current code, and there might be a bug where if a single chart is removed, all charts are uninstalled and re-installed. we should fix that if that's the case.

Reply:
per some internal discussions, we are going to modify the `getSortedCharts` method to use the kots HelmChart resources as the source for the deployments/undeployments rather than the presence of a directory